File: pkg_tags/version_0.7/lhs/R/maximinLHS.R
################################################################################
#
# Function: maximinLHS.R
# Purpose:  To create an optimum Latin Hypercube Sampling algorithm that is
#           implements the maximin criteria.  The maximin critera attemps to
#           maximize the minimum distance between points
# Author:   Rob Carnell
#           This program is motivated by the MATLAB program written by
#           John Burkardt
# Created:  May 05
#
# Discussion:
#    n Points in an k dimensional Latin hypercube are to be selected.
#    This algorithm tries to pick a solution which has the property that the
#    points are S optimal.
#    It uses the dup (DUPLICATION) factor to limit the number of points available
#    to choose from at each step.
#
#  Parameters:
#    k= positive integer, the spatial dimension.
#    n= positive integer, the number of points to be generated.
#    dup= positive integer, the DUPLICATION factor.
#      (default=1, a value of 5 is reasonable)
#
#  Steps
#   1.  Select a random starting point and place it in the matrix X
#   2.  Populate the matrix AVAIL with the integers from 1 to n
#   3.  Replace the values in AVAIL which have already been used by the
#       the first point in X with the value N
#   4.  Generate the valid points for each row in a random manner by using
#       the numbers in the rows of AVAIL
#   5.  For each cadidate point, calculate the distance to the points already
#       used in X.  Select the candidate point with the maximum minimum distance
#       value and place it in X (actually squared distance is used to eliminate
#       the square root processing step)
#   7.  Having chosen the new point for X, update AVAIL to replace those
#       numbers in each row that have been used.  The first "count" columns in
#       AVAIL are valid points.
#   8.  There is only one choice for the last point
#
################################################################################

maximinLHS <- function(n, k, dup=1)
{
  if(length(n)!=1 |length(k)!=1 | length(dup)!=1)
    stop("n, k, and dup may not be vectors")
  if(any(is.na(c(n,k,dup)))) stop("n, k, and dup may not be NA or NaN")
  if(any(is.infinite(c(n,k,dup)))) stop("n, k, and dup may not be infinite")
  if(n!=floor(n) | n<1) stop("n must be a positive integer\n")
  if(k!=floor(k) | k<1) stop("k must be a positive integer\n")
  if(dup!=floor(dup) | dup<1)
    stop("The dup factor must be a positive integer\n")

  result <- numeric(k*n)
  avail <- numeric(k*n)
  point1 <- numeric(k*(dup*(n-1)))
  list1 <- numeric(dup*(n-1))
  vec <- numeric(k)

  result2 <- .C("maximinLHS_C", as.integer(n), as.integer(k), as.integer(dup),
                as.integer(result), as.integer(avail), as.integer(point1),
                as.integer(list1), as.integer(vec))[[4]]

  eps <- runif(n*k)

  result2 <- (result2 - 1 + eps) / n

  return(t(matrix(result2, nrow=k, ncol=n, byrow=TRUE)))
}

File: src/maximinLHS.cpp
/**
 * @file maximinLHS.cpp
 * @author Robert Carnell
 * @copyright Copyright (c) 2014, Robert Carnell
 * 
 * License <a href="http://www.gnu.org/licenses/lgpl.html">GNU Lesser General Public License (LGPL v3)</a>
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "LHSCommonDefines.h"
#include "utilityLHS.h"

/*
 * Arrays are passed into this routine to allow R to allocate and deallocate
 * memory within the wrapper function.
 * The R internal random number generator is used that R can set.seed for
 * testing the functions.
 * Dimensions:  result  K x N
 * Parameters:
 *              N: The number of points to be sampled
 *              K: The number of dimensions (or variables) needed
 *              dup: The duplication factor which affects the number of points
 *                   that the optimization algorithm has to choose from
 * References:  Please see the package documentation
 *
 */
namespace lhslib 
{
    void maximinLHS(int n, int k, int dup, bclib::matrix<int> & result, bclib::CRandom<double> & oRandom)
    {
        if (n < 1 || k < 1 || dup < 1)
        {
            throw std::runtime_error("nsamples are less than 1 (n) or nparameters less than 1 (k) or duplication is less than 1");
        }
        msize_type nsamples = static_cast<msize_type>(n);
        msize_type nparameters = static_cast<msize_type>(k);
        unsigned int duplication = static_cast<unsigned int>(dup);
        if (result.rowsize() != nsamples || result.colsize() != nparameters)
        {
            throw std::runtime_error("result should be n x k for the lhslib::maximinLHS call");
        }
        result.transpose();
        // *****  matrix_unsafe<int> m_result = matrix_unsafe<int>(nparameters, nsamples, result);
        /* the length of the point1 columns and the list1 vector */
        msize_type len = duplication * (nsamples - 1);
        /* create memory space for computations */
        bclib::matrix<int> avail = bclib::matrix<int>(nparameters, nsamples);
        bclib::matrix<int> point1 = bclib::matrix<int>(nparameters, len);
        std::vector<int> list1 = std::vector<int>(len);
        std::vector<int> vec = std::vector<int>(nparameters);
        /* squared distance between corner (1,1,1,..) and (N,N,N,...) */
        double squaredDistanceBtwnCorners = static_cast<double>(nparameters * (nsamples - 1) * (nsamples - 1));

        /* index of the current candidate point */
        vsize_type point_index;
        /* index of the optimum point */
        unsigned int best;
        /* the squared distance between points */
        unsigned int distSquared;
        /* the minimum squared distance between points */
        double minSquaredDistBtwnPts;
        /*  The minumum candidate squared difference between points */
        unsigned int minCandidateSquaredDistBtwnPts;

        /* initialize the avail matrix */
        initializeAvailableMatrix(avail);

        /*
        * come up with an array of K integers from 1 to N randomly
        * and put them in the last column of result
        */
        for (msize_type irow = 0; irow < nparameters; irow++)
        {
            result(irow, nsamples-1) = static_cast<int>(std::floor(oRandom.getNextRandom() * static_cast<double>(nsamples) + 1.0));
        }

        /*
        * use the random integers from the last column of result to place an N value
        * randomly through the avail matrix
        */
        for (unsigned int irow = 0; irow < nparameters; irow++)
        {
            avail(irow, static_cast<unsigned int>(result(irow, nsamples - 1) - 1)) = static_cast<int>(nsamples);
        }

        /* move backwards through the result matrix columns */
        for (vsize_type ucount = nsamples - 1; ucount > 0; ucount--)
        {
            //unsigned int ucount = static_cast<unsigned int>(count);
            for (msize_type irow = 0; irow < nparameters; irow++)
            {
                for (msize_type jcol = 0; jcol < duplication; jcol++)
                {
                    /* create the list1 vector */
                    for (vsize_type j = 0; j < ucount; j++)
                    {
                        list1[j + ucount*jcol] = avail(irow, j);
                    }
                }
                /* create a set of points to choose from */
                for (msize_type jcol = ucount * duplication; jcol > 0; jcol--)
                {
                    point_index = static_cast<vsize_type>(std::floor(oRandom.getNextRandom() * static_cast<double>(jcol)));
                    point1(irow, jcol-1) = list1[point_index];
                    list1[point_index] = list1[jcol - 1];
                }
            }
            minSquaredDistBtwnPts = DBL_MIN;
            best = 0;
            for (msize_type jcol = 0; jcol < duplication * ucount - 1; jcol++)
            {
                /* set min candidate equal to the maximum distance to start */
                minCandidateSquaredDistBtwnPts = static_cast<unsigned int>(std::ceil(squaredDistanceBtwnCorners));
                for (msize_type j = ucount; j < nsamples; j++)
                {
                    distSquared = 0;
                    /*
                    * find the distance between candidate points and the points already
                    * in the sample
                    */
                    for (msize_type kindex= 0; kindex < nparameters; kindex++)
                    {
                        vec[kindex] = point1(kindex, jcol) - result(kindex, j);
                        distSquared += vec[kindex] * vec[kindex];
                    }
                    /*
                    * if the distance squared value is the smallest so far, place it in the
                    * min candidate
                    */
                    if (minCandidateSquaredDistBtwnPts > distSquared) 
                    {
                        minCandidateSquaredDistBtwnPts = distSquared;
                    }
                }
                /*
                * if the candidate point is the largest minimum distance between points so
                * far, then keep that point as the best.
                */
                if (static_cast<double>(minCandidateSquaredDistBtwnPts) > minSquaredDistBtwnPts)
                {
                    minSquaredDistBtwnPts = static_cast<double>(minCandidateSquaredDistBtwnPts);
                    best = static_cast<unsigned int>(jcol);
                }
            }

            /* take the best point out of point1 and place it in the result */
            for (msize_type irow = 0; irow < nparameters; irow++)
            {
                result(irow, ucount-1) = point1(irow, best);
            }
            /* update the numbers that are available for the future points */
            for (msize_type irow = 0; irow < nparameters; irow++)
            {
                for (msize_type jcol = 0; jcol < nsamples; jcol++)
                {
                    if (avail(irow, jcol) == result(irow, ucount-1))
                    {
                        avail(irow, jcol) = avail(irow, ucount-1);
                    }
                }
            }
        }

        /*
        * once all but the last points of result are filled in, there is only
        * one choice left
        */
        for (msize_type irow = 0; irow < nparameters; irow++)
        {
            result(irow, 0u) = avail(irow, 0u);
        }

        result.transpose();
    //#ifdef _DEBUG
        bool test = isValidLHS(result);

        if (!test)
        {
            /* the error function should send an error message through R */
            throw std::runtime_error("Invalid Hypercube\n"); // LCOV_EXCL_LINE
        }
    //#endif

    #if PRINT_RESULT
        lhsPrint(result, 0);
    #endif
    }
} // end namespace
